File: Heaps/Q9. Smallest Range in K Sorted List/manasvi9745.java
import java.util.*;

class Solution {
    public int[] smallestRange(List<List<Integer>> nums) {
        
        int k = nums.size();
        
        // Min heap: [value, listIndex, elementIndex]
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(
            (a, b) -> a[0] - b[0]
        );
        
        int currentMax = Integer.MIN_VALUE;
        
        // Initialize heap with first element from each list
        for (int i = 0; i < k; i++) {
            int val = nums.get(i).get(0);
            minHeap.offer(new int[]{val, i, 0});
            currentMax = Math.max(currentMax, val);
        }
        
        int rangeStart = 0;
        int rangeEnd = Integer.MAX_VALUE;
        
        while (true) {
            
            int[] curr = minHeap.poll();
            int minVal = curr[0];
            int listIndex = curr[1];
            int elemIndex = curr[2];
            
            // Update best range
            if (currentMax - minVal < rangeEnd - rangeStart) {
                rangeStart = minVal;
                rangeEnd = currentMax;
            }
            
            // Move to next element in same list
            if (elemIndex + 1 == nums.get(listIndex).size()) {
                break;  // One list exhausted
            }
            
            int nextVal = nums.get(listIndex).get(elemIndex + 1);
            minHeap.offer(new int[]{nextVal, listIndex, elemIndex + 1});
            
            currentMax = Math.max(currentMax, nextVal);
        }
        
        return new int[]{rangeStart, rangeEnd};
    }
}

File: Heaps/Q9. Smallest Range in K Sorted List/manasvi9745.py
import heapq
from typing import List

class Solution:
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        
        k = len(nums)
        heap = []
        current_max = float('-inf')
        
        # Initialize heap
        for i in range(k):
            value = nums[i][0]
            heapq.heappush(heap, (value, i, 0))  # (value, list_index, element_index)
            current_max = max(current_max, value)
        
        best_range = [-10**5, 10**5]
        
        while True:
            min_val, list_idx, elem_idx = heapq.heappop(heap)
            
            # Update best range
            if current_max - min_val < best_range[1] - best_range[0]:
                best_range = [min_val, current_max]
            
            # Move to next element in that list
            if elem_idx + 1 == len(nums[list_idx]):
                break  # One list exhausted
            
            next_val = nums[list_idx][elem_idx + 1]
            heapq.heappush(heap, (next_val, list_idx, elem_idx + 1))
            
            current_max = max(current_max, next_val)
        
        return best_range
